tensorzero-core/src/db/postgres/migrations/20250909185348_rate-limiting.sql

```diff
@@ -89,73 +90,212 @@ BEGIN
 )
-RETURNS TABLE (returned_tickets BIGINT)
+RETURNS TABLE (bucket_key_returned TEXT, final_balance BIGINT)
 LANGUAGE plpgsql AS $$
 DECLARE
-    bucket_state record;
+    i INT;
+    bucket_state RECORD;
     refilled_state calculated_bucket_state;
-    final_balance BIGINT;
+    temp_row RECORD;
+    new_balance BIGINT;
 BEGIN
-    -- Step 1: Atomically lock the row, creating it if it doesn't exist.
-    INSERT INTO resource_bucket (key, tickets, balance_as_of)
-    VALUES (p_key, p_capacity, now())
-    ON CONFLICT (key) DO UPDATE SET key = EXCLUDED.key
-    RETURNING * INTO bucket_state;
+    -- Input validation: check for consistent array lengths
+    IF array_length(p_keys, 1) IS NULL OR
+       array_length(p_keys, 1) != array_length(p_amounts, 1) OR
+       array_length(p_keys, 1) != array_length(p_capacities, 1) OR
+       array_length(p_keys, 1) != array_length(p_refill_amounts, 1) OR
+       array_length(p_keys, 1) != array_length(p_refill_intervals, 1) THEN
+        RAISE EXCEPTION 'Input arrays must have the same length';
+    END IF;
 
-    -- Step 2: Calculate the refilled state using the helper function.
-    SELECT * INTO refilled_state
-    FROM _calculate_refilled_state(
-        bucket_state.tickets,
-        bucket_state.balance_as_of,
-        p_capacity,
-        p_refill_amount,
-        p_refill_interval
-    );
+    -- Input validation: check for duplicate keys
+    IF array_length(p_keys, 1) > 0 AND array_length(p_keys, 1) != (
+        SELECT COUNT(DISTINCT key_column)
+        FROM unnest(p_keys) AS keys(key_column)
+    ) THEN
+        RAISE EXCEPTION 'Duplicate keys are not allowed in the input array';
+    END IF;
 
-    -- Step 3: Calculate the new balance, capped at capacity.
-    final_balance := least(
-        refilled_state.available_tickets + p_amount,
-        p_capacity
-    );
+    -- Create a temporary table to store inputs
+    CREATE TEMP TABLE temp_return_states (
+        key TEXT PRIMARY KEY,
+        amount BIGINT,
+        capacity BIGINT,
+        refill_amount BIGINT,
+        refill_interval INTERVAL
+    ) ON COMMIT DROP;
+
+    -- Populate temp table from input arrays
+    FOR i IN 1..array_length(p_keys, 1) LOOP
+        INSERT INTO temp_return_states (key, amount, capacity, refill_amount, refill_interval)
+        VALUES (p_keys[i], p_amounts[i], p_capacities[i], p_refill_amounts[i], p_refill_intervals[i]);
+    END LOOP;
+
+    -- Lock rows in a consistent order, calculate new state, and update.
+    FOR temp_row IN SELECT * FROM temp_return_states ORDER BY temp_return_states.key LOOP
+        -- Atomically get-or-create the bucket, locking the row.
+        INSERT INTO resource_bucket (key, tickets, balance_as_of)
+        VALUES (temp_row.key, temp_row.capacity, now())
+        ON CONFLICT (key) DO UPDATE SET key = EXCLUDED.key
+        RETURNING * INTO bucket_state;
+
+        -- Calculate the refilled state.
+        SELECT * INTO refilled_state
+        FROM _calculate_refilled_state(
```

Comment:
Can you make `_calculate_refilled_state` throw an exception if the interval is zero, rather than treating it as if no intervals have passed.

Reply:
yep, done + tested